### PR TITLE
perf(adapter): memoize the active bin + bin array per pool — RPC dedup (#211)

### PR DESCRIPTION
**Pull request by irfndi**
The performance audit's RPC-dedup — revalidated via codegraph (evaluatePool complexity 815) and direct code tracing before building.

## The waste (verified)
- `getActiveBin` fetched **twice per pool per cycle** — `getPoolState` (adapter-service.ts:2178) and `getBinArray` (:2220) both call the SDK's uncached active-bin fetch.
- `getBinsAroundActiveBin` (:2230) re-fetches the same bin array `getActiveBin` just loaded.
- ~7 RPC/pool/cycle → the doc's "100–200 RPS breaks standard tiers" is understated ~2× at 500 pools.

## The fix
A **short-TTL (3s) active-bin memo** shared by `getPoolState` + `getBinArray`:
- the within-cycle pair (milliseconds apart) costs ONE SDK fetch each;
- invalidated alongside the balance caches on every mutation (enter/exit/rebalance/claim);
- a later cycle past the TTL refetches fresh on-chain state — strictly tighter than the 5-min DLMM instance cache.

Net: ~3 RPC/pool/cycle removed (~7 → ~4), **zero behavior change**.

## Tests — 46 new (1729 total, +46)
- **2 memo proofs**: the pair shares one `getActiveBin` + one `getBinsAroundActiveBin` fetch; the memo expires past the 3s TTL on the same layer.
- **24 launch-gate boundaries**: age/TVL/volume/binStep/turnover exact edges, the 60s clock-skew tolerance, token-leg failures (unverified/stable/freeze), summarize slice + tie ordering, NaN/zero fee yields.
- **7 wash-forensics + 9 runner-dip thresholds**: exactly-at `WASH_MIN_TRADES`/unique-payer-rate/TPS, `stepPct` one-basis-point edges, max-steps, sizing caps, config-extreme dip anchors.
- **4 decision-loop edges** (3 subagents, files owned exclusively): no exit-and-reenter in one pass, timebox beats a runner scale-in, the wash gate rejects suspicious evidence, the launch-cap boundary.

Full suite 1729/1729; tsc 0; oxlint 0.

## Summary by Sourcery

Memoize DLMM active-bin and surrounding bin data per pool with a short TTL to reduce redundant RPCs while preserving behavior, and strengthen test coverage around launch gating, runner scale-in dip logic, wash-trade forensics thresholds, and decision-loop edge cases.

Tests:
- Add adapter active-bin memoization tests to verify shared SDK fetches across getPoolState/getBinArray and TTL expiry behavior.
- Extend launch-gate tests to cover boundary conditions for age, TVL, volume, binStep, turnover, token safety, fee-yield handling, and rejection summarization ordering and slicing.
- Expand runner dip and scale-in tests to validate dip anchor calculations, step triggers and limits, and top-up sizing constraints at configuration boundaries.
- Enhance wash-forensics tests to assert wash detection thresholds for trade count, unique payer rate, burst TPS, single-payer floors, and fee coefficient-of-variation computation.
- Add decision-loop tests to ensure no exit-and-reenter in a single pass, that timebox exits preempt runner scale-ins, that wash forensics can block unsafe launch ENTERs, and that launch-cap limits reject excess positions.

## Summary by CodeRabbit

* **Performance**
  * Reduced repeated pool-state and bin-array requests by reusing active-bin data for up to 3 seconds.
  * Keeps related liquidity data aligned and refreshes cached information after balance or position changes.
  * Prunes expired cached data automatically to maintain current results.

* **Tests**
  * Expanded coverage for launch eligibility, position limits, scale-in and top-up boundaries, wash-trading detection, fee calculations, edge cases, and memoized data requests.

### CODE REVIEW

File: engine/adapter-service.ts [hunk 1]
```diff
@@ -752,6 +753,65 @@ export const AdapterLive = Layer.effect(
...
+          Date.now() - memo.fetchedAt < ACTIVE_BIN_MEMO_TTL_MS
+        ) {
+          return memo.binsAround;
+        }
+        const bins = yield* Effect.tryPromise(() => dlmm.getBinsAroundActiveBin(20, 20));
+        activeBinMemo.set(poolAddress, {
+          binId: memo?.binId ?? bins.activeBin,
+          price: memo?.price ?? "",
+          fetchedAt: Date.now(),
+          binsAround: bins,
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:839]: Fixed: memoizedBinsAround now aligns the memo's binId/price with the bins fetch's OWN activeBin snapshot (the SDK reads the active bin internally) — a getBinArray just before expiry refreshes the memo with the fresh snapshot instead of extending the stale bin's life, and binId/price/binsAround always describe one point in time.

[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:839]: Fixed: the memo stores the bins fetch's own activeBin + its price (extracted from the matching bin), so the cached bins are always aligned with their active-bin snapshot — no mixing of an old id with newer bins, no stale-window extension.

File: bench/decision-loop-exit.test.ts [hunk 2]
```diff
@@ -919,3 +919,319 @@ describe("runner scale-in wiring (Heart Attack step 2)", () => {
...
+      configOverrides: {
+        watchlistPools: [POOL],
+        launchScanEnabled: true,
+        launchExecutionEnabled: true,
+        // The held $1k position is below the $2k dust threshold: the
+        // deterministic dust EXIT fires this cycle. The pool is ALSO a
+        // gate-passing launch candidate — with only this one position the
+        // exit frees the slot, so an unguarded ENTER slot would approve the
+        // re-entry; only the no-exit-and-reenter guard can stop it.
+        dustExitUsd: 2_000,
```
[the remark this answers is not in the document]
Comment (irfndi) [reply bench/decision-loop-exit.test.ts:1004]: Fixed: the no-exit-and-reenter test now passes datapi: makeHotDatapi(POOL) to the layer — without measured stats the ENTER candidate gate blocks every entry by design and the guard test passed vacuously.

File: engine/adapter-service.ts [hunk 3]
```diff
@@ -2217,18 +2285,17 @@ export const AdapterLive = Layer.effect(
       getBinArray: (poolAddress) =>
         Effect.gen(function* () {
           const dlmm = yield* getDlmm(poolAddress);
-          const activeBin = yield* Effect.tryPromise(() => dlmm.getActiveBin());
+          const activeBin = yield* memoizedActiveBin(poolAddress, dlmm);
           const halfRange = 20;
           const lowerBinId = activeBin.binId - halfRange;
           const upperBinId = activeBin.binId + halfRange;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:2342]: Fixed: getBinArray now derives its bounds, reported activeBinId, and the price-fallback base from realBins.activeBin (the fetch's own snapshot) instead of the local memo value — a pool moving between getPoolState and the bins fetch can no longer produce newer bins filtered with an older +-20 range while reporting reservesKnown: true. The local memo value remains only the null-bins fallback.

File: engine/adapter-service.ts [hunk 4]
```diff
@@ -752,6 +753,72 @@ export const AdapterLive = Layer.effect(
...
+    // going stale across cycles (the active bin moves with every swap). The
+    // DLMM instance itself is cached 5 min; this memo is strictly tighter.
+    const ACTIVE_BIN_MEMO_TTL_MS = 3_000;
+    interface ActiveBinMemo {
+      readonly binId: number;
+      readonly price: string;
+      readonly fetchedAt: number;
+      readonly binsAround: { activeBin: number; bins: BinLiquidity[] } | null;
+    }
+    const activeBinMemo = new Map<string, ActiveBinMemo>();
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:770]: Fixed: each memo read prunes entries aged past the TTL (opportunistic eviction) — a long-running paper agent's dropped pools no longer accumulate unreachable bin-array snapshots; the map stays bounded by the pools touched within one TTL window.

File: bench/adapter-active-bin-memo.test.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        layer,
+      ),
+    );
+    expect(dlmm.getActiveBin).toHaveBeenCalledTimes(1);
+
+    // Advance the clock past the 3s TTL on the SAME layer: the memo expires
+    // and the next read must see fresh on-chain state.
+    vi.spyOn(Date, "now").mockImplementation(() => realNow() + 5_000);
+    await Effect.runPromise(
+      Effect.provide(
```
[the remark this answers is not in the document]
Comment (irfndi) [reply bench/adapter-active-bin-memo.test.ts:176]: Fixed: the TTL test now advances a mutable clock INSIDE one provide — each read previously went through a fresh adapter lifetime (empty memo), so the expiry was never actually exercised. A second test covers the bins-memo half's expiry the same way.

File: engine/adapter-service.ts [hunk 6]
```diff
@@ -752,6 +753,88 @@ export const AdapterLive = Layer.effect(
...
+        // The bins fetch carries its OWN active-bin snapshot (the SDK reads
+        // the active bin internally): align the memo with it so binId/price
+        // and binsAround always describe the SAME point in time — retaining
+        // the old binId while refreshing fetchedAt would serve a stale active
+        // bin for another TTL window, and mixing bins from a newer snapshot
+        // with the old id would be internally inconsistent.
+        const binsActivePrice = bins.bins.find((b) => b.binId === bins.activeBin)?.price;
+        activeBinMemo.set(poolAddress, {
+          binId: bins.activeBin,
+          price: binsActivePrice ?? memo?.price ?? "",
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:836]: Fixed: an empty stored price (a bins-only fetch with a partial bin set and no prior memo) is now a MISS — the active-bin half refetches instead of serving currentPrice 0 from within the TTL window.

File: engine/adapter-service.ts [hunk 7]
```diff
@@ -752,6 +753,88 @@ export const AdapterLive = Layer.effect(
...
+        pruneActiveBinMemo();
+        const memo = activeBinMemo.get(poolAddress);
+        if (
+          memo &&
+          memo.binsAround !== null &&
+          Date.now() - memo.fetchedAt < ACTIVE_BIN_MEMO_TTL_MS
+        ) {
+          return memo.binsAround;
+        }
+        const bins = yield* Effect.tryPromise(() => dlmm.getBinsAroundActiveBin(20, 20));
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts]: Fixed: memoizedBinsAround now takes halfRange from getBinArray — the memo no longer hardcodes a ±20 window that could silently drop edge bins if the range ever widened.

File: bench/adapter-active-bin-memo.test.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        Effect.gen(function* () {
+          const adapter = yield* AdapterService;
+          yield* adapter.getPoolState(POOL_ADDRESS);
+        }),
+        layer,
+      ),
+    );
+    vi.restoreAllMocks();
+    expect(dlmm.getActiveBin).toHaveBeenCalledTimes(2);
```
[the remark this answers is not in the document]
Comment (irfndi) [reply bench/adapter-active-bin-memo.test.ts]: Fixed: the bins-memo half's TTL expiry is now tested (same-layer clock advance, getBinArray refetches past the TTL). The mutation-invalidation contract (invalidateBalanceCaches clears the memo) rides the existing balance-cache invalidation tested elsewhere — noted as the one seam still exercised only via the mutation paths.

File: engine/adapter-service.ts [hunk 9]
```diff
@@ -752,6 +753,91 @@ export const AdapterLive = Layer.effect(
...
+        pruneActiveBinMemo();
+        const memo = activeBinMemo.get(poolAddress);
+        if (memo && Date.now() - memo.fetchedAt < ACTIVE_BIN_MEMO_TTL_MS) {
+          return { binId: memo.binId, price: memo.price };
+        }
+        const activeBin = yield* Effect.tryPromise(() => dlmm.getActiveBin());
+        activeBinMemo.set(poolAddress, {
+          binId: activeBin.binId,
+          price: activeBin.price,
+          fetchedAt: Date.now(),
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:800]: Fixed: getPoolState restarts the memo's fetchedAt when the pool-state assembly completes — a cold/rotating pool whose metadata/reserves/pricing resolution exceeds 3s no longer has its memo entry pruned by the immediately following getBinArray, so the dedup holds in the high-latency, high-cardinality case it exists for.

File: engine/adapter-service.ts [hunk 10]
```diff
@@ -752,6 +753,91 @@ export const AdapterLive = Layer.effect(
...
+          binsAround: null,
+        });
+        return { binId: activeBin.binId, price: activeBin.price };
+      });
+    }
+
+    function memoizedBinsAround(
+      poolAddress: string,
+      dlmm: DLMM,
+      halfRange: number,
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:811]: Fixed: the memo entry now carries the halfRange it was fetched with and the bins hit path requires a match — a future wider-range call site would miss instead of being served the narrower cached window.

File: engine/adapter-service.ts [hunk 11]
```diff
@@ -2175,14 +2266,25 @@ export const AdapterLive = Layer.effect(
...
 
+          // The TTL clock starts when the pool-state ASSEMBLY completes, not
+          // when the active bin was fetched: for a cold or rotating pool the
+          // metadata/reserves/pricing resolution can exceed 3s, and the
+          // immediately following getBinArray call must still hit the memo
+          // (otherwise the dedup is defeated exactly in the high-latency,
+          // high-cardinality scenario it exists for).
+          const assembledMemo = activeBinMemo.get(poolAddress);
+          if (assembledMemo) {
+            activeBinMemo.set(poolAddress, { ...assembledMemo, fetchedAt: Date.now() });
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts]: Fixed: the assembly-end re-stamp now applies only when this getPoolState actually FETCHED the active bin (memo miss) — a screener-populated entry (bins fetched by a standalone getBinArray) is never refreshed past its TTL by a later getPoolState.

[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts]: Fixed (WARNING): the re-stamp is hit-guarded — only a call that actually fetched re-stamps after assembly; hits leave fetchedAt untouched, so freshness stays a hard 3s cap rather than caller-cadence-dependent.

File: bench/adapter-active-bin-memo.test.ts [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+          // Advance the clock past the 3s TTL: the next read must refetch.
+          fakeNow += 5_000;
+          yield* adapter.getPoolState(POOL_ADDRESS);
+        }),
+        layer,
+      ),
+    );
+    vi.restoreAllMocks();
+    expect(dlmm.getActiveBin).toHaveBeenCalledTimes(2);
```
[the remark this answers is not in the document]
Comment (irfndi) [reply bench/adapter-active-bin-memo.test.ts:165]: Addressed: the TTL re-stamp was hit-guarded so it cannot extend a hit entry's life or refresh older screener-populated bins; the memo tests cover the hit + miss + expiry paths.
